File: 0587-erect-the-fence/0587-erect-the-fence.js
/**
 * @param {number[][]} trees
 * @return {number[][]}
 */
var outerTrees = function(trees) {
  trees.sort((a, b) => a[0] === b[0] ? a[1] - b[1] : a[0] - b[0]);
  const lower = [];
  const upper = [];
  
  for (let tree of trees) {
    while (lower.length >= 2 && cmp(lower.at(-2), lower.at(-1), tree) > 0) lower.pop();
    while (upper.length >= 2 && cmp(upper.at(-2), upper.at(-1), tree) < 0) upper.pop();
    
    lower.push(tree);
    upper.push(tree);
  }
  
  return [...new Set([...upper, ...lower])];
};

const cmp = (p1, p2, p3) => {
  const [x1, y1] = p1;
  const [x2, y2] = p2;
  const [x3, y3] = p3;
  
  return (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1);
}

File: 0587-erect-the-fence/0587-erect-the-fence.py
class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        def cmp(p1, p2, p3):
            x1, y1 = p1
            x2, y2 = p2
            x3, y3 = p3
            return (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1)
        
        
        trees.sort()
        lower = []
        upper = []
        for tree in trees:
            while len(lower) >= 2 and cmp(lower[-2], lower[-1], tree) > 0:
                lower.pop()
            while len(upper) >= 2 and cmp(upper[-2], upper[-1], tree) < 0:
                upper.pop()
            lower.append(tuple(tree))
            upper.append(tuple(tree))
        return list(set(lower + upper))